### PR TITLE
feat: DNS-01 challenge support

### PR DESCRIPTION
**Pull request by buchdag**
This PR pick up the work done by @daweedm in #1022 ( thanks 👍 )

I tested it with Gandi Live DNS, seems to work without issue.

@pini-gh I know you worked on a DNS-01 fork too, so a review would be welcome.

Ideally I'd also like this to include wildcard certificate compatibility before merge.

### CODE REVIEW

**Comment by buchdag** [comment]
Added support for wildcard certificates.

**Comment by buchdag** [comment]
This is available as `nginxproxy/acme-companion:dns01` on Dockerhub for anyone who want to try it.

**Comment by pini-gh** [comment]
LGTM.

**Comment by pgassmann** [comment]
Can you support configuring the ACMESH_DNS_API_CONFIG on the acme-companion container instead of the proxied application?
Usually one DNS API config per proxy should be enough and in that way, the api token is not available to the proxied application.
You coud add a DEFAULT_ACMESH_DNS_API_CONFIG, like the DEFAULT_EMAIL

**Comment by buchdag** [comment]
@pgassmann done, updated the `nginxproxy/acme-companion:dns01` image, will update docs later.

**Comment by buchdag** [comment]
I don't think this can be tested in CI without obtaining a certificate from a real CA (let's say from Let's Encrypt staging environment) for a real domain whose DNS can be manipulated by acme.sh DNS API. I have a few ideas how to do this but I think it'll need some changes to the test system setup itself before, so this feature will probably ship without CI tests at first.

File: app/start.sh [hunk 1]
```diff
@@ -7,7 +7,9 @@ term_handler() {
 
     # shellcheck source=functions.sh
     source /app/functions.sh
-    remove_all_location_configurations
+    if parse_true "${ACME_HTTP_CHALLENGE_LOCATION:=false}"; then
+        remove_all_location_configurations
+    fi
```
Comment (buchdag) [inline app/start.sh:12]: I think this cause some intermittent CI failure. We'll tackle this in #1128 or later.
```suggestion
    remove_all_location_configurations
```

File: README.md [hunk 2]
```diff
@@ -10,20 +10,24 @@ It handles the automated creation, renewal and use of SSL certificates for proxi
...
-* Check your firewall rules and [**do not attempt to block port `80`**](https://letsencrypt.org/docs/allow-port-80/) as that will prevent `http-01` challenges from completing.
+* Check your firewall rules and [**do not attempt to block port `80`**](https://letsencrypt.org/docs/allow-port-80/) as that will prevent `HTTP-01` challenges from completing.
 * For the same reason, you can't use nginx-proxy's [`HTTPS_METHOD=nohttp`](https://github.com/nginx-proxy/nginx-proxy#how-ssl-support-works).
 * The (sub)domains you want to issue certificates for must correctly resolve to the host.
-* Your DNS provider must [answer correctly to CAA record requests](https://letsencrypt.org/docs/caa/).
 * If your (sub)domains have AAAA records set, the host must be publicly reachable over IPv6 on port `80` and `443`.
 
+If you can't meet these requirements, you can use the `DNS-01` challenge instead. Please refer to the [documentation](./docs/Let's-Encrypt-and-ACME.md#dns-01-acme-challenge) for more information.
```
Comment (buchdag) [inline README.md:27]: Use absolute URL
```suggestion
If you can't meet these requirements, you can use the `DNS-01` challenge instead. Please refer to the [documentation](https://github.com/nginx-proxy/acme-companion/blob/main/docs/Let's-Encrypt-and-ACME.md#dns-01-acme-challenge) for more information.
```

**Comment by buchdag** [comment]
If there isn't any further comment or review on this, I'll think I'll merge on Sunday or Monday.

**Comment by cxhamilton** [comment]
I tested this feature branch on a real host.domain.tld on Namecheap and the DNS challenge worked perfectly. 

Then, I changed LETSENCRYPT_HOST in my container's YAML from host.domain.tld to *.domain.tld. It also worked, but I'm getting an SSL error when I try to access https://host.domain.tld.

Is there something I'm doing wrong from a wildcard domain perspective?

**Comment by pini-gh** [comment]
> I'm getting an SSL error

Would you mind giving a few more details?

**Comment by cxhamilton** [comment]
I was trying to provide detailed documentation, going step-by-step to reproduce the problem, but **I can no longer do so.**

I've also added a second service using the same wildcard value for LETSENCRYPT_HOST and it **works as expected**.

For reference, here is a relevant sample from a docker compose file:

```
services:
    heimdall:
        container_name: heimdall
        image: lscr.io/linuxserver/heimdall:latest
        environment:
            - PUID=${PUID}
            - PGID=${PGID}
            - TZ=${TZ}
            - VIRTUAL_HOST=heimdall.mydomain.com
            - LETSENCRYPT_HOST=*.mydomain.com
```

The certs that are being installed by acme-companion:

```
[Sun Jul 28 14:04:24 UTC 2024] Installing cert to: /etc/nginx/certs/wildcard_mydomain.com/cert.pem
[Sun Jul 28 14:04:24 UTC 2024] Installing CA to: /etc/nginx/certs/wildcard_mydomain.com/chain.pem
[Sun Jul 28 14:04:24 UTC 2024] Installing key to: /etc/nginx/certs/wildcard_mydomain.com/key.pem
[Sun Jul 28 14:04:24 UTC 2024] Installing full chain to: /etc/nginx/certs/wildcard_mydomain.com/fullchain.pem
```

At this point, I assume that it was a mistake that I made in the configuration.

**Pull request merged**